tests_common/deflake.py

```diff
@@ -0,0 +1,341 @@
+from __future__ import annotations
+
+import os
+from dataclasses import dataclass, field
+from pathlib import Path
+from subprocess import run, PIPE
+from typing import Generator, cast
+
+import pytest
+import requests
+from _pytest import runner
+from _pytest.config import Config, Parser
+from _pytest.main import Session
+from _pytest.nodes import Item
+from _pytest.reports import TestReport
+from _pytest.runner import CallInfo
+from _pytest.stash import StashKey
+from _pytest.terminal import TerminalReporter
+
+TEST_TYPE_OPTION = "--deflake-test-type"
+PREVIOUS_OUTCOME_KEY = StashKey[dict[str, str]]()
+
+APP_REPO = "snowflakedb/snowflake-cli"
+ISSUE_REPO = APP_REPO
+FLAKY_LABEL = "flaky-test"
+PHASES = ["setup", "call", "teardown"]
+
+
+class DeflakePlugin:
+    """Pytest plugin to retry tests and mark them as flaky.
+
+    This plugin is used to determine which tests are flaky by automatically
+    retrying them and marking them as "flaky" instead of "passing" if they
+    pass on the retry. After the test run, a GitHub issue is opened or updated
+    for each flaky test encountered.
+    """
+
+    name = "deflake"
+
+    def __init__(self: DeflakePlugin, config: Config) -> None:
+        super().__init__()
+        self.exceptions: dict[str, Exception] = {}
+        self.flaky_tests = 0
+        self.test_run = TestRun()
+
+        self.runner: type[runner] = config.pluginmanager.getplugin("runner")
+        self.test_type: str = config.getoption(TEST_TYPE_OPTION)
+
+        if token := os.getenv("GH_TOKEN"):
+            # Grab the current commit so we can generate absolute URLs
+            rev_parse = run(["git", "rev-parse", "HEAD"], text=True, stdout=PIPE)
+            self.github: GitHub | None = GitHub(token, rev_parse.stdout.strip())
```

Comment:
Have you considered running `gh` in subprocess? If the only thing we want to do is playing with issues then `gh issue edit`

https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/using-github-cli-in-workflows

In this way we may reduce complexity by removing `GitHub` class. And it's a bit safer - the authentication is handled by Github.

Reply:
i'd still like to have the `GitHub` class to encapsulate all the issue manipulation, and I find it cleaner to deal with HTTP APIs when practical over spawning subprocesses. In terms of security, the GH_TOKEN still needs to be in an env var even when using the `gh` CLI, so I don't think it makes a big difference. If you feel strongly about it I can change it